server/src/main/java/org/elasticsearch/action/admin/indices/alias/TransportIndicesAliasesAction.java

```diff
@@ -110,16 +111,26 @@ protected void masterOperation(Task task, final IndicesAliasesRequest request, f
         //Expand the indices names
         List<AliasActions> actions = request.aliasActions();
         List<AliasAction> finalActions = new ArrayList<>();
-
         // Resolve all the AliasActions into AliasAction instances and gather all the aliases
         Set<String> aliases = new HashSet<>();
         for (AliasActions action : actions) {
-            final Index[] concreteIndices = indexNameExpressionResolver.concreteIndices(state, request.indicesOptions(), action.indices());
+            final Index[] concreteIndices = indexNameExpressionResolver.concreteIndices(state, request.indicesOptions(), false,
+                action.indices());
+            for (Index concreteIndex : concreteIndices) {
+                IndexAbstraction indexAbstraction = state.metadata().getIndicesLookup().get(concreteIndex.getName());
+                assert indexAbstraction != null : "invalid cluster metadata. index [" + concreteIndex.getName() + "] was not found";
+                if (indexAbstraction.getParentDataStream() != null) {
+                    throw new IllegalArgumentException("The provided expressions [" + String.join(",", action.indices())
+                        + "] match a backing index belonging to data stream [" + indexAbstraction.getParentDataStream().getName()
+                        + "]. data streams and their backing indices don't support aliases.");
```

Comment:
Minor: We're inconsistent across the codebase in whether we capitalize our error messages, but I think we should be consistent within a single error message. In other words, either make "The provided..." lowercase or capitalize the second sentence:

```suggestion
                        + "]. Data streams and their backing indices don't support aliases.");
```